### ISSUE TITLE
[imports] HTML Includes (<import src='header.html'/> in <body>)

### ISSUE THREAD

**Issue by techtonik** [WICG/webcomponents#280]
The requirement to use javascript to insert HTML snippet into main HTML document is daunting. The behavior that imported HTML is stored in strangely accessible way (through `link[import]` selector) is mystic. The user expectation is to have:

```
<body>
  <import src='header.html'/>
</body>
```

**Comment by hayatoito** [comment]
I understand the expectation, however, you can't use HTML Imports for the purpose because HTML Imports are not designed to support such the expectation.

**Comment by hayatoito** [comment]
BTW, I think it's not so difficult to make a custom element which meets your expectation, e.g. `<include-here src="hearder.html">`, by using the primitives provided by Custom Elements and HTML Imports.

**Comment by waldyrious** [comment]
Just adding that the expectation is a quite reasonable one. Quoting [this article](http://www.html5rocks.com/en/tutorials/webcomponents/imports/):

> Think about how you load different types of resources on the web. For JS, we have `<script src>`. For CSS, your go-to is probably `<link rel="stylesheet">`. For images it's `<img>`. Video has `<video>`. Audio, `<audio>`.... Get to the point! The majority of the web's content has a simple and _declarative_ way to load itself. Not so for HTML. (...) See the irony? **The web's most basic content, HTML, requires the greatest amount of effort to work with.**

For now the closest we can get to that is [seamless iframes](https://html.spec.whatwg.org/#attr-iframe-seamless), but the `seamless` attribute still needs to be emulated using CSS until it becomes [supported by the browsers](http://caniuse.com/#feat=iframe-seamless).

**Comment by techtonik** [comment]
> I understand the expectation, however, you can't use HTML Imports for the
> purpose because HTML Imports are not designed to support such the
> expectation.
> 
> So, is it possible to (re)design it to match the expectation? I'd even
> extend it for CSS and SVG.

BTW, how do you deal with complexity of issues? Do you maintain a graph of
use cases connected to possible solutions, which are in turn connected to
problems that come with each solution. Github is too primitive to maintain
such a system, so maybe there is a parallel tracker?

**Comment by domenic** [comment]
What you are proposing is essentially an entirely separate spec. I would suggest that if you're interested in making it happen you go through [all the usual steps for proposing a new feature for the platform](https://wiki.whatwg.org/wiki/FAQ#Is_there_a_process_for_adding_new_features_to_a_specification.3F). And, while you're doing so, be sure to think up a new name, since HTML imports is already taken for a different feature.

**Comment by techtonik** [comment]
Creating a separate spec even called `HTML Includes` with `<include>` or `<import>` tag will be totally confusing when there is already `HTML Imports` spec. But why importing HTML in the `<body>` falls out of spec for `HTML Imports`?

If it is not desirable to introduce new tag, it might be possible to further hack `<link rel>` and allow it inside of `<body>`.

**Comment by chris-l** [comment]
One way to see it, is that `HTML Imports` work more similar to a node.js `require`; in the sense that the solicited content is like a module that is now available to be used and manipulated - but you need to actually write code that uses it!

Instead, what you are proposing is more similar to a PHP `include`; the way it works is like if the line doing the `include` is just replaced by the raw content of the imported file.

It falls out of spec, because `HTML Imports` was designed for the former scenario, (that is, having the content "in scope", ready to be manipulated by the coder), and not for inserting raw content.

Is not that is undesirable creating a new tag, is just that `HTML Import` is not about including raw content, so doing that would be the work of a completely new, different spec.

**Comment by techtonik** [comment]
I don't know what is node.js `require`.

**Comment by techtonik** [comment]
Still I don't see the answer to the question why raw HTML `include` in the body can not be made the part of `HTML Import` spec? Where is the point of conflict?

**Comment by annevk** [comment]
Includes and imports are fundamentally different. It wouldn't make sense to conflate the two.

**Comment by waldyrious** [comment]
> Still I don't see the answer to the question why HTML Import in the body can not be made the part of this spec?

@techtonik as much as I agree with you that there should be such a functionality, I have to say that I don't see any clearer way to put this than what @chris-l has said above. Perhaps your lack of familiarity with `require` is the barrier here. These are fundamentally different concepts:
- **HTML imports** is similar to Node.js' `require` or Python's `import`, in that it gets stuff ready to be used in your code. Sort of like placing tools in your toolbox to be in hands' reach. (I am not entirely sure if importing, say, CSS resources doesn't apply them immediately, which might be confusing in this regard.)
- A hypothetical **HTML includes** functionality would function like PHP's `include`, directly inserting HTML content into the current document, like iframe currently does but more integrated (think `<img>`, `<video>`, etc). I'd totally back a proposal for such a spec, which as mentioned in the passage I quoted above, is sorely lacking in HTML.

@chris-l any thoughts on whether such a spec could/should be part of the Web Components family?

**Comment by techtonik** [comment]
@waldyrious thanks for the explanation. That clarifies things. I am not ready yet to propose a **HTML Include** spec (with `<include>` instead of `<link>` hack). Is there an official web component to "polyfill" this functionality in the meanwhile?

**Comment by techtonik** [comment]
My primary use case for now is to include common header:

```
<include src="header.html"/>
```

This should work locally without server side software.

This raises security issues (as with PHP). It may happen that pages in subdirs need to include stuff from parent dirs, and this may lead to a lot of security issues. If I understand correctly, right now locally saved HTML file with malicious JS still, while loaded with the browser, still can not access local filesystem and read sensitive files.

**Comment by chris-l** [comment]
@waldyrious If `HTML Includes` should be part of Web Components?
I don't think so. While that functionality would indeed be useful, the other specs from Web Components are more directly related to isolation.

All this features are related with "isolation", while an `include`, is almost the opposite (the content should be _transparently_ included, that is, the included content should be part of the DOM, just like if was in the same HTML file and no extra/special commands should be necessary to access those nodes)

An `HTML Include` could be used inside or outside Web components, just like `div`, `script` or any other element. So I don't really see it as an spec that fits inside of Web components, IMHO.

(as side note, `HTML Include` must be **blocking**, for actually allowing full transparency)

**Comment by waldyrious** [comment]
@chris-l thanks, that confirms my initial guess.

@techtonik as I mentioned above, the way this can currently be done is using iframes. The seamless attribute should be enough to make it work for simple cases, but currently [it isn't implemented](http://caniuse.com/#feat=iframe-seamless) in any of the major browsers, so you have to emulate it manually with CSS. As an example, you can see the iframe approach in use in my [personal website](http://waldyrious.net) (see [here](https://github.com/waldyrious/waldyrious.github.io/blob/36ffdd1ce5933235eff74a7a6471bea8de2a41a1/index.xhtml#L14) and [here](https://github.com/waldyrious/waldyrious.github.io/blob/36ffdd1ce5933235eff74a7a6471bea8de2a41a1/css/style.css#L30) for the relevant code).

**Comment by techtonik** [comment]
@waldyrious something tells me that this won't be implemented. Also, how can I highlight current item in `<iframe>` for my `header.html` navigation?

**Comment by waldyrious** [comment]
> @waldyrious something tells me that this won't be implemented.

Agreed that this doesn't seem to be a priority for now. But the expectation of declarative inclusion of external documents is natural enough (imo) to lead me to believe we might get something along those lines in the future.

> Also, how can I highlight current item in `<iframe>` for my header.html navigation?

I believe a `:focus` pseudoselector should work, but that discussion is off-topic on this issue. Please use another forum for that.

**Comment by chris-l** [comment]
@techtonik Well, since this seemed like an interesting idea, I decided to create a web component that implement this: [html-include](https://github.com/chris-l/html-include).

And you can use it the way you wanted - inside body:

``` html
<html-include src="header.html"></html-include>
```

That will load the `header.html` file with a `XMLHttpRequest` and replace itself with the raw content.
It will stop the propagation of the normal `DOMContentLoaded` event, and once all the `html-include` elements are resolved, it will then manually emit a `DOMContentLoaded` event, so the listeners it could have see the fully composed DOM.

**Comment by techtonik** [comment]
@chris-l very nice. I will try it. I'd definitely miss the `DOMContentLoaded` part if I ever get to implementing web components myself.

If I understand it correctly, it still requires local web server to work, but at least there are no security issues. Is that right?

**Comment by chris-l** [comment]
@techtonik To not hijack this thread, I created an [issue](https://github.com/chris-l/html-include/issues/1) on it's issues section. But the short answer, is no. No more than doing ajax.

**Comment by tomalec** [comment]
@techtonik I have implemented a custom element, that uses HTML Imports to achieve what (I think) you need https://github.com/Juicy/imported-template

It uses HTML Import to load the document (Import nested HTML Imports/dependencies, executes a scripts, etc) then stamps a Document Fragment from it, to your wherever you put `<template is="imported-template">`.

I have just made an article on how it works http://starcounter.io/html-partialsincludes-webcomponents-way/

To sum up my thoughts on original issue, I would say: 

Having HTML Imports, Templates, and Custom Elements, we do not need a spec for what "HTML Includes" we just need a small convention.
